### PR TITLE
Fix spurious error when a `Drop` local has an assignment in a loop

### LINKED ISSUE #70919: spurious "borrow might be used .. when [variable] is dropped and runs the destructor"
[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ca0e9f5171bf484d9b744d08057203ec)
```rust
use std::rc::Rc;
use std::cell::{RefCell, RefMut};
use std::mem::drop;

fn huh(x: Rc<RefCell<i32>>) {
    let mut rc: Rc<RefCell<i32>> = x.clone();
    let mut inner: RefMut<'_, i32> = rc.borrow_mut();
    loop {
        drop(inner);
        drop(rc);
        rc = x.clone();
        inner = rc.borrow_mut();
    }
}
```
produces:
```
error[E0505]: cannot move out of `rc` because it is borrowed
  --> src/lib.rs:9:14
   |
6  |     let mut inner: RefMut<'_, i32> = rc.borrow_mut();
   |                                      -- borrow of `rc` occurs here
...
9  |         drop(rc);
   |              ^^ move out of `rc` occurs here
...
12 |         inner = rc.borrow_mut();
   |         ----- borrow might be used here, when `inner` is dropped and runs the destructor for type `std::cell::RefMut<'_, i32>`

error[E0506]: cannot assign to `rc` because it is borrowed
  --> src/lib.rs:11:9
   |
6  |     let mut inner: RefMut<'_, i32> = rc.borrow_mut();
   |                                      -- borrow of `rc` occurs here
...
11 |         rc = x.clone();
   |         ^^ assignment to borrowed `rc` occurs here
12 |         inner = rc.borrow_mut();
   |         ----- borrow might be used here, when `inner` is dropped and runs the destructor for type `std::cell::RefMut<'_, i32>`
```

(Same error with `-Zpolonius`.)

The error messages are wrong: the assignment to `inner` can never run a destructor, since the previous value of `inner` was dropped earlier on.

And AFAICT this code is sound and ought to be accepted.

### PR DESCRIPTION
**Pull request by Aaron1011**
Fixes https://github.com/rust-lang/rust/issues/70919

`DropAndReplace` terminators are special - unlike all other terminators, they perform two distinct actions (a drop and a store) to the same `Place`. This complicates various analysis passes that we do, which expect at most one of 'Drop'/'Def'/'Use' for a local at a given MIR location.

Previously, we categorized a `DropAndReplace` terminator's `Local` access as `MutatingUseContext::Drop`. As a result, livness computation would not consider the `DropAndReplace` as a store ('Def') of a local. The "use live" set would end up being too large (a use would be seen to extend back to an earlier store, rather than the closure `DropAndReplace`).

We now explicitly propagate information about `DropAndReplace` terminators via new enum variants `MutatingUseContext:DropAndReplace` and `DefUse::DropAndReplace`. We use this information in liveness computation to record *both* a Drop and a Def. This prevents creating an extra-large "use live" set, while ensuring that te local is still considered "drop live" at the required locations.

### CODE REVIEW

File: compiler/rustc_mir_dataflow/src/impls/liveness.rs [hunk 1]
```diff
@@ -179,6 +179,9 @@ impl DefUse {
                 }
             }
 
+            // This terminator writes a new value, so it's a Def
+            PlaceContext::MutatingUse(MutatingUseContext::DropAndReplace) => Some(DefUse::Def),
```
Comment (tmiasko) [inline compiler/rustc_mir_dataflow/src/impls/liveness.rs:183]: I would expect to find this case together with arm above, i.e., it is a def but only when it is a direct place without projections. Anyway, is this actually reachable?

**Comment by Aaron1011** [comment]
@tmiasko Updated

File: compiler/rustc_mir_dataflow/src/impls/liveness.rs (hunk 1 above)
Comment (tmiasko) [reply compiler/rustc_mir_dataflow/src/impls/liveness.rs]: Ah, my suggestion wasn't quite right either. The drop precedes the assignment in the evaluation order, so even if there were a definition, the overall effect would be a use of a local regardless.

**Comment by Aaron1011** [comment]
After doing additional testing, I've found that this PR has an odd effect on this code:

```rust
struct HasDrop<T>(T);
impl<T> Drop for HasDrop<T> {
    fn drop(&mut self) {}
}

#[allow(unused_assignments)]
#[allow(unused_variables)]
fn foo() {
    let mut first: u8 = 0;
    let mut has_drop = Some(HasDrop(&first));

    loop {
        match true {
            true => {
                has_drop = Some(HasDrop(&first));
            }
            false => {
                drop(has_drop);
                has_drop = None;
                let reborrow = &mut first;
            }
        }
        
    }
}

fn main() {}
```

It compiles, but *only* if you have *both* `drop(has_drop)` and `hash_drop = None`. I think the handling of `DropAndReplace` in `ConstraintGeneration` is doing something odd (it doesn't go through `visit_local`).

I'm getting worried that this change is actually unsound in some subtle way. If I can't figure out a simpler way of doing this, then I think we should try to get rid of `DropAndReplace` entirely, as discussed in https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/.60DROP.60.20to.20.60DROP_IF.60.20compiler-team.23558.

**Pull request closed without merging**